### ISSUE TITLE
Relocation against symbol can not be used when making a shared object

### ISSUE THREAD

**Issue by krytarowski** [dotnet/coreclr#3020]
```
[ 98%] Building CXX object src/jit/dll/CMakeFiles/ClrJit.dir/__/lowerxarch.cpp.o
[ 98%] Building CXX object src/jit/dll/CMakeFiles/ClrJit.dir/__/codegenxarch.cpp.o
[ 98%] Building CXX object src/jit/dll/CMakeFiles/ClrJit.dir/__/simdcodegenxarch.cpp.o
[ 98%] Building CXX object src/jit/dll/CMakeFiles/ClrJit.dir/__/simd.cpp.o
[ 98%] Building CXX object src/jit/standalone/CMakeFiles/ryujit.dir/__/lower.cpp.o
[ 98%] Building CXX object src/jit/standalone/CMakeFiles/ryujit.dir/__/lsra.cpp.o
[ 98%] Building CXX object src/jit/standalone/CMakeFiles/ryujit.dir/__/targetamd64.cpp.o
[ 98%] Building CXX object src/jit/standalone/CMakeFiles/ryujit.dir/__/unwindamd64.cpp.o
[ 98%] Building CXX object src/jit/standalone/CMakeFiles/ryujit.dir/__/emitxarch.cpp.o
[ 98%] Building CXX object src/jit/standalone/CMakeFiles/ryujit.dir/__/lowerxarch.cpp.o
[ 98%] Building CXX object src/jit/standalone/CMakeFiles/ryujit.dir/__/codegenxarch.cpp.o
[ 98%] Building CXX object src/jit/standalone/CMakeFiles/ryujit.dir/__/simdcodegenxarch.cpp.o
[ 98%] Building CXX object src/jit/standalone/CMakeFiles/ryujit.dir/__/simd.cpp.o
[ 98%] Linking CXX static library libClrJit.a
[ 98%] Built target ClrJit
Scanning dependencies of target coreclr
[ 98%] Building CXX object src/dlls/mscoree/coreclr/CMakeFiles/coreclr.dir/__/mscoree.cpp.o
[ 98%] Building CXX object src/dlls/mscoree/coreclr/CMakeFiles/coreclr.dir/__/unixinterface.cpp.o
[ 98%] Linking CXX shared library libryujit.so
clang-3.9: warning: argument unused during compilation: '-nostdinc++'
clang-3.9: warning: argument unused during compilation: '-g'
[ 98%] Linking CXX shared library libcoreclr.so
clang-3.9: warning: argument unused during compilation: '-nostdinc++'
clang-3.9: warning: argument unused during compilation: '-g'
[ 98%] Built target ryujit
/usr/bin/ld: ../../../vm/wks/libcee_wks.a(externalmethodfixupthunk.S.o): relocation R_X86_64_PC32 against symbol `ExternalMethodFixupWorker' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value
clang-3.9: error: linker command failed with exit code 1 (use -v to see invocation)
--- src/dlls/mscoree/coreclr/libcoreclr.so ---
```

```
$ pkg_info |grep -E 'llvm|lldb|clang'
llvm-3.9.0nb20160131 Low Level Virtual Machine compiler infrastructure
clang-3.9.0nb20160131 C language family frontend for LLVM
lldb-3.9.0nb20160131 next generation, high-performance debugger
```

```
$ uname -a
NetBSD chieftec 7.99.25 NetBSD 7.99.25 (GENERIC) #0: Fri Dec 25 20:51:06 UTC 2015  root@chieftec:/tmp/netbsd-tmp/sys/arch/amd64/compile/GENERIC amd64
```

How should it be addressed?

Regarding `-nostdintc++`, I'm using Clang with `libstdc++`, not `libc++`.

**Comment by jkotas** [comment]
Many different projects hit [this error message](http://www.bing.com/search?q=relocation+R_X86_64_PC32+against+symbol+can+not+be+used+when+making+a+shared+object%3B+recompile+with+-fPIC). The fix seems to always be a tweak to the compiler or linker command line options.

**Comment by krytarowski** [comment]
I was suggested that there is missing `-fPIC` somewhere.

**Comment by jasonwilliams200OK** [comment]
Either `-fPIC` is missing (which it doesn't seem like after going through every CMakeLists) or it needs to come immediately after `clang` for the version of clang you are using. Atleast this used to be a requirement/bug in `gcc` not very long ago.

**Comment by krytarowski** [comment]
Can you try out with clang from trunk on Linux or FreeBSD?

**Comment by krytarowski** [comment]
Assuming that `pkgsrc` is in `/usr/pkgsrc` and `pkgsrc-wip` in `/usr/pkgsrc/wip`:

```
cd /usr/pkgsrc/wip/coreclr-git && bmake
```

**Comment by jasonwilliams200OK** [comment]
I used clang35 from Q42015 ports to build CoreCLR on FreeBSD: https://gist.github.com/jasonwilliams200OK/06db121ec534821fe541.

Also since the CoreFX's native code-base is small, I tried building it with clang37 and succeeded: https://gist.github.com/jasonwilliams200OK/6efa7907e66275df2d24.

I am configuring NetBSD 7 x86_64 in VM on Mac at the moment (don't have a charger with me though :disappointed:). OS is installed, using pkg_add to install cmake atm.

Will see if i could help figuring out this and other issues you have reported.
You may also want to drop by https://gitter.im/dotnet/coreclr channel sometimes. Other people might able to suggest quicker solutions/hacks.

**Comment by krytarowski** [comment]
Thank you! I suggest doing it through source-based pkgsrc. The framework will feed CoreCLR with needed dependencies and paths.

**Comment by krytarowski** [comment]
But... You can use pkgsrc on Linux, FreeBSD, MacOSX, Cygwin etc.

**Comment by jasonwilliams200OK** [comment]
@krytarowski, working with NetBSD compared to FreeBSD is bit inconvenient.

> python: no module named pyexpact

Took me sometime to resolve the python package dependency `pyexpat`, which:

1. does not get installed even with `pkg_add -u` and `pkgin -F` reinstall commands; although it is python's core module.
2. listed as py27-expat, py34-expat and so on in `pkgin` search list (yeah, so now i am using `pkgin` instead of pkg_add and pkgsrc :smile:). This un-bolocked me (and this turned out to be one of those ungoogle-able problem.. as most Linux issues are resolved by reinstalling the package ..)

This is for reference for the time when the repo CI will be configured for NetBSD, @mmitche  and @ellismg might need this info. Since mine is a fresh install, I will keep the steps documented on gist for future reference.

So clang36, python34 and 27 (one of them is required though), cmake33 are installed. Cmake configuration fails for libunwind though and my mac's battery is almost dead. I will continue later. Meanwhile, would be very helpful if you could push your changes in a separate branch and share the link for disabling the libunwind checks in #2910. Thanks!

**Comment by krytarowski** [comment]
Thank you for your effort!

Please go for http://pkgsrc.org/wip/ and try to build coreclr-git. Currently there are ~20 patches disabling this or that.

I (re)started a NetBSD port of libunwind. It will take a while!

**Comment by krytarowski** [comment]
I will add the `py-expat` dependency to `pkgsrc-wip/coreclr-git`

**Comment by jasonwilliams200OK** [comment]
Official build is great idea! :+1: 

I will add the py-expat dependency to pkgsrc-wip/coreclr-git

`py27-expat` to be exact (as NetBSD7 by default ships with python-2.7).

**Comment by krytarowski** [comment]
NetBSD doesn't ship with any python version (so far). pkgsrc is a separated project, detached from NetBSD - and with different release cycle.

```
diff --git a/coreclr-git/Makefile b/coreclr-git/Makefile
index 77c2aee..bc75dc5 100644
--- a/coreclr-git/Makefile
+++ b/coreclr-git/Makefile
@@ -18,6 +18,8 @@ USE_LANGUAGES=        c c++
 USE_TOOLS+=    bash
 REPLACE_BASH+= build.sh
 
+DEPENDS+=      ${PYPKGPREFIX}-expat-[0-9]*:../../textproc/py-expat
+
 do-build:
                ${RUN} cd ${WRKSRC} && PYTHON=${PYTHONBIN} ./build.sh
 
```

**Comment by krytarowski** [comment]
If you really want to build CoreCLR out of pkgsrc, here are patches:

https://wip.pkgsrc.org/cgi-bin/gitweb.cgi?p=pkgsrc-wip.git;a=tree;f=coreclr-git/patches;hb=HEAD

However this is untested.

I don't use Gitter, I sit on #netbsd-pl @ FreeNode. There are more people interested in the .Net Core porting to NetBSD. Feel free to join.

**Comment by jasonwilliams200OK** [comment]
Thanks @krytarowski! I ramble around at #FreeBSD, #node and #libuv from time to time. I will drop by. :)

**Comment by krytarowski** [comment]
It will be great, I have got many general questions.

**Comment by krytarowski** [comment]
This has been fixed in pkgsrc-wip https://wip.pkgsrc.org/cgi-bin/gitweb.cgi?p=pkgsrc-wip.git;a=commitdiff;h=63c5b19c6edfc12e7b9ee2e3eaa59e59a741c509

I will cleanup and upstream this today.

**Comment by jasonwilliams200OK** [comment]
@krytarowski, congrats! I just read your PTL bit, but not sure where it is going to get incorporated. Will watch out for you PR. :+1:

**Comment by krytarowski** [comment]
Closed via #3149
